Source/WebCore/dom/ScriptExecutionContext.h

```diff
@@ -331,6 +332,27 @@ class ScriptExecutionContext : public SecurityContext, public TimerAlignment {
     void addDeferredPromise(Ref<DeferredPromise>&&);
     RefPtr<DeferredPromise> takeDeferredPromise(DeferredPromise*);
 
+    template<typename Promise, typename Task>
+    void enqueueTaskWhenSettled(Ref<Promise>&& promise, TaskSource taskSource, Task&& task)
+    {
+        // FIXME: To remove the NativePromiseRequest allocation, we could take the NativePromiseCallback from the whenSettled value.
+        auto request = makeUnique<NativePromiseRequest>();
+        WeakPtr weakRequest { *request };
+        auto command = promise->whenSettled(nativePromiseDispatcher(), [weakThis = WeakPtr { *this }, taskSource, task = WTFMove(task), request = WTFMove(request)] (auto&& result) mutable {
+            request->complete();
+            RefPtr protectedThis = weakThis.get();
+            if (!protectedThis)
+                return;
+            protectedThis->eventLoop().queueTask(taskSource, [task = WTFMove(task), result = WTFMove(result)] () mutable {
+                task(WTFMove(result));
+            });
+        });
+        if (weakRequest) {
```

Comment:
this is rather convoluted. Something seems amiss.

if nativePromiseDispatcher() is a dispatching to a different thread that would cause weakRequest to become null, then the use of WeakPtr isn't thread-safe. 

If it is the same thread, then weakRequest will never be null and the test is unnecessary.